Send confirmation of registration by email

brucebolt opened the pull request:
What
----

Sends an email when registration has been completed. provided an email address has been provided.

Code has been copied (mostly verbatim) from https://github.com/alphagov/govuk-coronavirus-business-volunteer-form.

Not to be merged until:
- email content and subject line has been confirmed
- environment variables have been set for the PaaS app
- Redis is correctly linked for the PaaS app

<img width="656" alt="Screenshot 2020-05-04 at 12 22 02" src="https://user-images.githubusercontent.com/6329861/80961015-ef5a3780-8e01-11ea-82a4-2029b9a18789.png">

How to review
-------------

- Review code and tests
- Run application locally to ensure user flow through the check your answers page has not been affected

Links
-----

Trello card: https://trello.com/c/JDy5EWw4

brucebolt committed:
Add Sidekiq

This will be used for queuing the sending of confirmation emails and SMS
messages.

brucebolt committed:
Add content for confirmation email

brucebolt committed:
Add GOV.UK Notify Ruby client

brucebolt committed:
Add email job and mailer

These are copied from govuk-coronavirus-business-volunteer-form and will
allow us to send confirmation emails using GOV.UK Notify.

brucebolt committed:
Send a confirmation email on registration

If a user provides an email address, we will send an email when they
submit the form.

brucebolt committed:
Include GOV.UK Notify environment variables in Concourse pipeline

These variables will allow us to define the Notify API key and template
ID.

bilbof commented on app/controllers/coronavirus_form/check_answers_controller.rb:
```diff
@@ -25,6 +25,8 @@ def submit
         FormResponse: session,
       )
+
+      send_confirmation_email if session.dig(:contact_details, :email).present?
```
I deployed this in a second step for the other app, but we're more confident this time I don't see why shouldn't do it all in one step

bilbof commented on concourse/tasks/deploy-to-govuk-paas.yml:
```diff
@@ -51,6 +53,8 @@ run:
       cf set-env govuk-coronavirus-vulnerable-people-form GA_VIEW_ID "$GA_VIEW_ID"
+      cf set-env govuk-coronavirus-business-volunteer-form GOVUK_NOTIFY_TEMPLATE_ID "$GOVUK_NOTIFY_TEMPLATE_ID"
+      cf set-env govuk-coronavirus-business-volunteer-form NOTIFY_API_KEY "$NOTIFY_API_KEY"
 
```
I think you need to set the `PAAS_ENV` here, since you use that here https://github.com/alphagov/govuk-coronavirus-vulnerable-people-form/pull/296/commits/a9c75bd572bf3da01b9c1eca45c8e64b5897bb33#diff-ebeb2548534e3e5e7c407258f3d531bdR59

brucebolt replied:
Good spot.  Have added this.

bilbof reviewed (commented):
Looks good. Just looks like `PAAS_ENV` needs adding

bilbof reviewed (dismissed):
Looks good 👍

brucebolt committed:
Always use Notify for email delivery

Use GOV.UK Notify for sending emails, rather than defaulting to SMTP.

brucebolt committed:
Add Redis service to PaaS application manifest

brucebolt committed:
Scale worker instances on deployment

brucebolt committed:
Add mailers queue to Sidekiq configuration

brucebolt committed:
Update text about email confirmation

We are now sending confirmation emails, so need to update the text that
says we will not send an email on the confirmation page

bilbof commented on concourse/pipeline.yml:
```diff
@@ -136,6 +139,7 @@ jobs:
         params:
           CF_SPACE: production
           INSTANCES: 20
+          WORKER_INSTANCES: 1
```
Would be cool to keep an eye on this, since this application is fairly high traffic, whether we'd need more worker instances..

There's a card on our planning board to capture sidekiq stats in our prometheus metrics, which should help.

brucebolt replied:
Yeah, I'll keep an eye on this for the next few days

bilbof reviewed (approved):
Nice work! 🎉
